announcement blogpost (#24)
biscuit-auth/website: content/blog/biscuit-2.0.md

+++
title = "Biscuit 2.0 release"
description = "Introducing Biscuit 2.0"
date = 2022-01-16T09:19:42+00:00
updated = 2022-14-01T09:19:42+00:00
draft = true
template = "blog/page.html"

[taxonomies]
authors = ["Rustaceans"]

[extra]
lead = "Biscuit 2.0 announcement"
+++

We are delighted to announce the release of Biscuit at version 2.0! Biscuit is a specification
for a cryptographically verified authorization token supporting offline delegation, and a
language for authorization policies based on Datalog. It has been in development for 3 years
and is already deployed in production systems.

Biscuit was designed to address new challenges in building authorization in distributed systems:
- a query going through microservices must be verifiable decentrally without giving its full

> authrization level to every node

Reviewer: ```suggestion
authorization level to every node
```

- API authorization increasingly has to fit a user's authorization landscape (teams, roles,
delegation) while guaranteeing multitenant isolation

As an example, [Clever Cloud](https://www.clever-cloud.com)'s Pulsar service is accessible

> with Biscuit tokens: each customer is provided a token with full access to their namespace,

Reviewer: ```suggestion
with Biscuit tokens: each customer is given a token with full access to their namespace,
```

(else it would be "provided with", which would be a bit redundant)

and from there new tokens can be created that are limited to producing or subscribing to
specific topics, exactly according to application needs, without adding specific code on
the server's side.

To support those use cases, Biscuit specifies a token with public key signatures (like JWT)
and offline attenuation (like Macaroons): any service that knows the public key can verify
the token. And any token holder can derive a new token with more restrictions. For the
microservices use case, that means a node can take the token it received, and create a new one

> with the least privilege before sending it to the next node.

Reviewer: ```suggestion
with fewer privileges before sending it to the next node.
```

To guarantee that authorization policies execute the same way everywhere, Biscuit provides
an authorization language based on Datalog, that all implementations must support.
This logic language comes with benefits about execution (guaranteed to terminate) and
serialization (small enough to be carried in the token) while keeping nearly as much

> expressiveness as SQL.

Reviewer: "expressiveness" as in "easy to read", or as in "able to encode specific things"? If that's the latter, one could argue that SQL is _a bit more_ expressive.

Author: more the latter, and I added "nearly as much" because we're not there yet (we'd still need aggregated queries and a few other things). Readability sounds good

You can test it right there, in your browser:

> {% datalog() %}

Reviewer: (and yes, I know i should provide a way for token-printer to take datalog blocks in addition to an encoded token)

> {% datalog() %}

Author: right, that would be great! I just copied that example there, but there could be a better one

> {% datalog() %}

Reviewer: I'll craft a token when I have time

// we receive a request to read "admin.doc"
// The request contains a token with the following content
user(1234);

> // this restricts the kind of operation to "read"

Reviewer: the token itself restricts its uses to only `read` operations.

check if operation("read");

// The authorizer loads facts representing the request
resource("admin.txt");
operation("read");

// The authorizer loads the user's rights
right(1234, "admin.txt", "read");
right(1234, "admin.txt", "write");

// Finally, the authorizer tests policies
// by looking for a set of facts matching them
allow if
  user($user_id),
  resource($res),
  operation($op),
  right($user_id, $res, $op);
{% end %}

While Biscuit has great ambitions for your systems, it can be integrated right
now without replacing the entire authorization stack: you can use it to carry

> user ids and API keys, and benefit from attenuation on the client side.

Reviewer: maybe an example would be nice, a token carrying a session id, with stateless checks (TTL, and source ip address for instance)

The second change is about Datalog execution. In a token, the first block contains
the initial rights as facts, created by the root of trust. In 1.0, to avoid confusion
with facts from the next blocks, they were tagged with the `#authority` symbol, and
facts provided by the authorizer from request data had the `#ambient` tag.

In 2.0, Datalog execution is better isolated, it makes sure that there will be no
interference from later blocks, without requiring `#authority` or `#ambient`.
This simplifies writing policies significantly.

We also removed entirely the Symbol type (marked with the `#`). Symbols were interned
strings, separated from normal strings. They were used to reduce the token's size: if
a symbol appeared multiple times, the token would only carry the string once, and refer
to it by a number. It also improved performance of Datalog execution, because symbols
could be matched by comparing numbers instead of string equality. It came with a
tradeoff: symbols did not support string operations like prefix matching.
Now all strings are interned, supporting all operations, so the symbols are not needed
anymore, and execution gets a performance boost.

> 

Reviewer: ```suggestion
In addition to spec changes, there are new implementations and tooling available:

- a [haskell implementation](https://hackage.haskell.org/package/biscuit-haskell) covers all the v2 spec and comes with [bindings for protecting servant endpoints](https://hackage.haskell.org/package/biscuit-servant)
- the [wasm implementation](https://www.npmjs.com/package/@biscuit-auth/biscuit-wasm) makes the library usable from NodeJS and browsers, with both CommonJS an ES6 modules, as well as typescript definitions
- [web components](https://www.npmjs.com/package/@biscuit-auth/web-components) provide a simple way to interact with biscuits client-side (see for instance the datalog playground and the token inspector used on the website)
```

> Outline:

Reviewer: ```suggestion
While Rust, JS and Haskell implementations fully support v2.0 biscuits, there is still work to be done:

[biscuit-go](todo), [biscuit-java](todo), [biscuit-C#](todo) and [biscuit-swift](todo) don't support V2 yet and we welcome help getting them there.
We would also like to improve documentation and examples, so if you want to discuss use cases, or find the existing material unclear, please reach out so we can improve it!

Finally, the big breaking changes (serialization, cryptographic schemes) have been shipped in V2, but there are still open questions about new features, namely providing PKI primitives within datalog, finding a way to encode something similar to macaroon's third-party caveats in biscuit, and extending biscuit to support specific ecdsa profiles, in order to make hardware tokens support easier.

Outline:
```

with facts from the next blocks, they were tagged with the `#authority` symbol, and
facts provided by the authorizer from request data had the `#ambient` tag.

In 2.0, Datalog execution is better isolated, it makes sure that there will be no
interference from later blocks, without requiring `#authority` or `#ambient`.
This simplifies writing policies significantly.

### Removing the Symbol type

We also removed entirely the Symbol type (marked with the `#`). Symbols were interned
strings, separated from normal strings. They were used to reduce the token's size: if
a symbol appeared multiple times, the token would only carry the string once, and refer
to it by a number. It also improved performance of Datalog execution, because symbols
could be matched by comparing numbers instead of string equality. It came with a
tradeoff: symbols did not support string operations like prefix matching.
Now all strings are interned, supporting all operations, so the symbols are not needed
anymore, and execution gets a performance boost.

### New implementations

In addition to spec changes, there are new implementations and tooling available:

> - a [command line application](https://github.com/biscuit-auth/biscuit-cli) to create, inspect and attenuate tokens

Reviewer: ```suggestion
- a [command line application](https://github.com/biscuit-auth/biscuit-cli) to create, inspect, authorize and attenuate tokens
```

## Come help us!

While Rust, JS and Haskell implementations fully support v2.0 biscuits, there is still work to be done:

[biscuit-go](https://github.com/biscuit-auth/biscuit-go), [biscuit-java](https://github.com/clevercloud/biscuit-java), [biscuit-dotnet](https://github.com/fbredy/biscuit-dotnet) and [biscuit-swift](https://github.com/RemiBardon/biscuit-swift) don't support V2 yet and we welcome help getting them there.
We would also like to improve documentation and examples, so if you want to discuss use cases, or find the existing material unclear, please reach out so we can improve it! Come [chat with us](https://matrix.to/#/#biscuit-auth:matrix.org).

Finally, the big breaking changes (serialization, cryptographic schemes) have been shipped in V2, but there are still open questions about new features, namely providing PKI primitives within datalog, finding a way to encode something similar to macaroon's third-party caveats in biscuit, and extending biscuit to support specific ecdsa profiles, in order to make hardware tokens support easier.

> 

Reviewer: ```suggestion

While biscuit v1 was the consolidation of the initial ideas, biscuit v2 is the result of production experience and external feedback. We are super happy to see biscuit become more mature and excited for it to find new use cases.
```

A first attempt at a conclusion. With that I think it's fit for release
